examples/reference/custom_components/ReactiveESM.md

```diff
@@ -0,0 +1,336 @@
+        python_datetime = datetime.datetime.fromtimestamp(unix_timestamp)
+        self.value = str(python_datetime)
+
+button = ButtonEventExample()
+pn.Column(
+    button, button.param.value,
+).servable()
+```
+
+## Dependency Imports
+
+JavaScript dependencies can be directly imported via URLs, such as those from [`esm.sh`](https://esm.sh/).
+
+```python
+import panel as pn
+
+pn.extension()
+
+class ConfettiButton(pn.ReactiveESM):
+
+    _esm = """
+    import confetti from "https://esm.sh/canvas-confetti@1.6.0";
+
+    export function render() {
+        let btn = document.createElement("button");
+        btn.innerHTML = `Click Me`;
+        btn.addEventListener("click", () => {
+            confetti()
+        });
+        return btn
+    }
+    """
+
+ConfettiButton().servable()
+```
+
+Use the `_import_map` attribute for more concise module references.
+
+```python
+import panel as pn
+
+pn.extension()
+
+class ConfettiButton(pn.ReactiveESM):
+    _import_map = {
+        "canvas-confetti": "https://esm.sh/canvas-confetti@1.6.0",
+    }
+
+    _esm = """
+    import confetti from "canvas-confetti";
+
+    export function render() {
+        let btn = document.createElement("button");
+        btn.innerHTML = `Click Me`;
+        btn.addEventListener("click", () => {
+            confetti()
+        });
+        return btn
+    }
+    """
+
+ConfettiButton().servable()
+```
+
+## External Files
+
+You can load JavaScript and CSS from files by providing the paths to these files.
+
+Create the file **counter_button.py**.
+
+```python
+from pathlib import Path
+
+import param
+import panel as pn
+
+pn.extension()
+
+class CounterButton(pn.ReactiveESM):
+
+    value = param.Integer()
+
+    _esm = Path("counter_button.js")
+    _stylesheets = [Path("counter_button.css")]
+
+CounterButton().servable()
+```
+
+Now create the file **counter_button.js**.
+
+```javascript
+export function render({ data }) {
+    let btn = document.createElement("button");
+    btn.innerHTML = `count is ${data.value}`;
+    btn.addEventListener("click", () => {
+        data.value += 1;
+    });
+    data.watch(() => {
+        btn.innerHTML = `count is ${data.value}`;
+    }, 'value');
+    return btn;
+}
+```
+
+Now create the file **counter_button.css**.
+
+```css
+button {
+    background: #0072B5;
+    color: white;
+    border: none;
+    padding: 10px;
+    border-radius: 4px;
+}
+button:hover {
+    background: #4099da;
+}
+```
+
+Serve the app with `panel serve counter_button.py --autoreload`.
+
+You can now edit the JavaScript or CSS file, and the changes will be automatically reloaded.
+
+- Try changing the `innerHTML` from `count is ${data.value}` to `Count is ${data.value}` and observe the update.
+- Try changing the background color from `#0072B5` to `#008080`.
+
+## React with JSX
+
+To use [React](https://react.dev/) with [JSX](https://react.dev/learn/writing-markup-with-jsx) instead of plain JavaScript, follow this approach:
```

Comment:
I think we should consider whether support for React should be 

- Automatic (as now)
- Optional (via `_framework="js"` or `_framework="react"`)
- In a seperate `ReactComponent` component
- Documented

I have experienced a few hard to reproduce issues when changing code from vanilla js to React and back when testing the examples. Its just much more complicated to automatically support the combination and it might also slow things down. We should also test if other popular frameworks vue or svelte works and use that as input for decission.

Reply:
My vote would be to either leave it as is OR have distinct `JSComponent` and `ReactComponent` (leaving the door open for `VueComponent` etc.) types. In terms of performance there's really no difference, thanks to the compiler it's trivial to detect whether JSX is used and therefore switch to React but once Vue is used I'm not sure it's that easy.